### ISSUE TITLE
KT1025 / BT201

### ISSUE THREAD

**Issue by sbridger** [christian-kramer/JieLi-AC690X-Familiarization#4]
I assume that this is a custom firmware with BT (and usb) audio + serial. 
A very useful combo. 
Its [datasheet](https://nettips.ru/book/BT201%20Module%20_KT1025A_B_User%20Manual%20_V2.3.pdf) has some info about updating the firmware.

I am still trying to find the document about serial referred to in the datasheet: 
"26_Detailed instructions on Bluetooth data transmission and AT instructions. pdf" 
(Added file in case link disappears)
[BT201 Module _KT1025A_B_User Manual _V2.3.pdf](https://github.com/christian-kramer/JieLi-AC690X-Familiarization/files/15049965/BT201.Module._KT1025A_B_User.Manual._V2.3.pdf)

**Comment by das-Sourav** [comment]
YOU CAN CONNECT IT THROUGH ARDUINO FOR AT COMMENTS

**Comment by christian-kramer** [comment]
> YOU CAN CONNECT IT THROUGH ARDUINO FOR AT COMMENTS

Do you mean AT Commands?

**Comment by das-Sourav** [comment]
> > YOU CAN CONNECT IT THROUGH ARDUINO FOR AT COMMENTS
> 
> Do you mean AT Commands?

Yes sorry about my spelling mistakes

**Comment by christian-kramer** [comment]
> > > YOU CAN CONNECT IT THROUGH ARDUINO FOR AT COMMENTS
> > 
> > 
> > Do you mean AT Commands?
> 
> Yes sorry about my spelling mistakes

No problem at all, I was just making sure I understood. Do you have more information about how to use AT Commands with AC690X chips? I didn't know they had support for this.

**Comment by sbridger** [comment]
I assume that KT1025A is a custom firmware for the AC690X chip, and that the AT commands are a feature of that firmware.

My question is: has anyone found the document referred to.

It is a pretty useful firmware, at least for me as I need both  audio in/out and a serial control channel.

**Comment by KhairanarBharat** [comment]
Did anyone found the "updata.bfu" file somewhere?

**Comment by sottoalberi** [comment]
The official Jieli updater tool has a serial interface to get onto the device but the AT command set is a mystery ..for experimentation. Start with just AT+ or AT help. KT1025A = AC6925A

The .bfu file should be in the post build folder of the SDK. It will be a specific bfu file based on how the chip has been configured by the original designer. They are basically the full .bin file loaded into the device.

https://www.aliexpress.com/item/1005004688181549.html

**Comment by sottoalberi** [comment]
> > > > YOU CAN CONNECT IT THROUGH ARDUINO FOR AT COMMENTS
> > > 
> > > 
> > > Do you mean AT Commands?
> > 
> > 
> > Yes sorry about my spelling mistakes
> 
> No problem at all, I was just making sure I understood. Do you have more information about how to use AT Commands with AC690X chips? I didn't know they had support for this.

Check this from GitHub re using AT commands on BT201 module  https://youtu.be/ca2di44stXk 
 https://github.com/MooreSound/code/blob/master/BT201_Simple.ino

**Comment by timetofly06** [comment]
Hi everyone !

Searched for hours and didn't find anywhere the last 2.3 firmware for the BT201 module. Can someone send the updata.bfu file to me or post the link to it ? 

Thanks you so much guys

**Comment by moss-maurice** [comment]
Section 13 of the documentation states that you can customize sounds. You need to place the BT.mp3 and other files in the root of the TF card. The file size should not exceed 17kb. I did so, but the files are not customizable. Tell me, what am I doing wrong?

**Comment by Victossom** [comment]
Hello, good morning, has anyone managed to change the Bluetooth name?

**Comment by hakangucyetmez** [comment]
> Hello, good morning, has anyone managed to change the Bluetooth name?

Yes I did , use lightblue and the adress is fff3 to write the at'bdxxxx   xxx is the new name then reboot.First connect to spp ssid

dou uou have uddata.bfu file ples I must flash my board...

**Comment by hakangucyetmez** [comment]
I need updata.bfu file please send to me hakan.gucyetmez@hotmail.com

**Comment by Akashsujanapal** [comment]
I need the file pls mail akashsujanapal1977@gmail.com

**Comment by uastudiopro** [comment]
I made such a program
https://www.youtube.com/shorts/Q_M81AoFDSg
Mobile auto answer and it helps to login, where a login it's a number of phone

**Comment by PhuocBull** [comment]
Hi fellas, I'm also building a Bluetooth speaker using this module.
I found the latest (assuming, **V2.4 / 2020-04-17**) firmware **[here](http://www.szqyvhome.com/DownLoad/18685.html)**.
Just put the "updata.bfu" file in USB/SD card (FAT/FAT32, 32GB max), connect the power, wait 10s or something, it's best to hook the module with a 3-5w speaker or amplifier with speakers because you will hear few beeping sound then it stops and the update process is done.
"updata.bfu" file location : 20240524154050.zip/99_BT201方案_KT1025A_B_资料_V2.4_20200426/02_用户手册/

**Comment by umitmorgul** [comment]
@PhuocBull Hello do you know that can we put a limit of the top volume level ?

**Comment by PhuocBull** [comment]
> @PhuocBull Hello do you know that can we put a limit of the top volume level ?

I really don't know what you mean by that.
Anyway, this module has 31 volume levels, from 0 to 30, and you can set the volume using push buttons or serial commands. The last volume value will be saved even when powering off. One volume level applies across all modes. For example, if you set the volume to 20 in Bluetooth mode, after switching to USB, it will still be 20.

**Comment by hakangucyetmez** [comment]
Hi did you test the new firmware mr ümit ? is there any difference about volume limiting ? also any other diffrence on new firmware ?

hakan
________________________________
Gönderen: BLVDE ***@***.***>
Gönderildi: 1 Ekim 2024 Salı 17:14
Kime: christian-kramer/JieLi-AC690X-Familiarization ***@***.***>
Bilgi: hakangucyetmez ***@***.***>; Comment ***@***.***>
Konu: Re: [christian-kramer/JieLi-AC690X-Familiarization] KT1025 / BT201 (Issue #4)


@PhuocBull<https://github.com/PhuocBull> Hello do you know that can we put a limit of the top volume level ?

I really don't know what you mean by that.
Anyway, this module has 31 volume levels, from 0 to 30, and you can set the volume using push buttons or serial commands. The last volume value will be saved even when powering off. One volume level applies across all modes. For example, if you set the volume to 20 in Bluetooth mode, after switching to USB, it will still be 20.

**Comment by umitmorgul** [comment]
deneme şansı bulamadım şehir dışındayım ancak dökümanlara biraz baktım ve
limitlemeye dair bir şey göremedim

5 Eki 2024 Cmt, saat 22:27 tarihinde hakangucyetmez <
***@***.***> şunu yazdı:

**Comment by PhuocBull** [comment]
My BT201 on the V2.4 firmware.
So technically, nothing had changed. All the serial commands are still the same, and everything else.
31 levels of volume (0 to 30) aren't the problem, guys, It still  goes 0 to 100%. They made it in 31 levels so users don't have to click so many times just to adjust the volume. 0 is 0%, 30 is 100%, get it ?

If you're making a bluetooth speaker with B201, I have a few advices :
1. Don't use the onboard amp, obviously, it's suck, it's a single channel, it's only 3-5W, connect the onboard 3.5mm to some amplifier board with some cables.
2. If you want to add the AUX input feature, use an audio relay, switch between BT201 3.5mm and AUX input to the amp
3. Search for AT+CK in the PDF, turn on bluetooth running in background, it'll be useful (more on this below)
4. When your BT201 already turned on bluetooth in backgound, it'll still send feedback about bluetooth status to serial while in other modes like USB/TF, PC sound card. You can take advanted of this, put some audio files in an SD card and use AT+AF, AT+AJ (READ THE PDF !!!) to play the these audio files as Prompt Tones (AT+CN00 to turn off onboard tones)

I don't have much free time, but will make a complete guide on this module when the project is done.

![image](https://github.com/user-attachments/assets/e58f6f83-9029-45a7-92ab-aeb10c3b9496)

**Comment by PhuocBull** [comment]
> > > > > YOU CAN CONNECT IT THROUGH ARDUINO FOR AT COMMENTS
> > > > 
> > > > 
> > > > Do you mean AT Commands?
> > > 
> > > 
> > > Yes sorry about my spelling mistakes
> > 
> > 
> > No problem at all, I was just making sure I understood. Do you have more information about how to use AT Commands with AC690X chips? I didn't know they had support for this.
> 
> Check this from GitHub re using AT commands on BT201 module https://youtu.be/ca2di44stXk https://github.com/MooreSound/code/blob/master/BT201_Simple.ino

By the way, fellas, check out his github, very useful to explore this board, it has so much potential.

**Comment by sbridger** [comment]
> ... while in other modes like ..., PC sound card.

Does it have USB sound card mode? 
Does it also have USB-> serial?

If it did, it would be incredibly useful!

**Comment by PhuocBull** [comment]
Yes, usb sound card is available, use a male to male cable and plug it in your pc.
USB -> Serial is uncertain.
In USB sound card mode, the name appears to be BT201-AUD1O -BLE and I don't know how to change this name yet, regular Bluetooth and BLE Bluetooth's name are changeable as usual
![image](https://github.com/user-attachments/assets/c24fc327-0f56-482e-a46a-3ba9cc8b1411)

**Comment by Svatochlup** [comment]
Hi, I have tried the MooreSound repo. It is awesome, But i cant figure out how to rename the chip name displayed in bluetooth settings in Phone. I tried writing this code:

```
#include <SoftwareSerial.h>

SoftwareSerial BT201Serial(4, 3); // RX, TX


int btStatus = 0;

unsigned long previousMillis = 0; 
char BLEname;


void resetBT201(){
  BT201Serial.println("AT+CZ");
   delay(100);
}

int getBTStatus(){
String readStatus;
 BT201Serial.println("AT+TS");
  delay(50);
   while (BT201Serial.available() > 0) {
    char cs = BT201Serial.read();
    if (cs == '+') {
      if (readStatus.length() >0) {
        Serial.println(readStatus); //prints string to serial port out

        int ns = readStatus.toInt();  //convert readString into a number
        if(ns == 2 || ns == 4){
      return 3;
    }else if(ns == 1){
     return 2;
     }else if(ns == 3){
     return 4;
    }else if(ns == 0){
      return 1;
    }

        readStatus=""; //clears variable for new input
      }
    }  
    else {     
      readStatus += cs; //makes the string readString
    }
   
}
}

void setup() {
 Serial.begin(115200);
 BT201Serial.begin(115200);
    resetBT201();
    btStatus = getBTStatus();
    delay(300);

 BT201Serial.println("AT+BDLOSFAKOSrn");
 delay(500);
 resetBT201();
 delay(3000);
 Serial.println("Renamed done");
 
}

void loop() {

}

```
But it didnt changed the name. I hopeless. If anyone could help me I would appreciate it. Thanks a lot!
@PhuocBull

**Comment by PhuocBull** [comment]
@Svatochlup Sorry, late reply. It should be : BT201Serial.println("AT+BDLOSFAKOS"); .
Changing Bluetooth's name doesn't change a device's MAC. So try again, after rename your module, go to your phone Bluetooth setting, remove the BT201 from your phone's connected devices list, search again, it should work. And you only need to change the name once,  BT201 will save it.
Tip : If you don't have a USB to TTL serial adapter to perform a simple rename command or baseline setting on your BT201, you can use an Arduino nano/UNO, etc. Connect (short) GND pin to RST on the Arduino, now you have one. Process to connect Arduino's RX pin to RX and TX to TX on BT201. Yes RX to RX and TX to TX. Plug the Arduino board to your computer, select the port.

**Comment by budimanaabuu** [comment]
Has anyone ever successfully used the AT command AT+AJ/01*/011_11.????
I would like to specify "01???" as the folder inside "011_11", and "???" as the file to be played once.

Similarly, using AT+AJ/02*/011*.???, the idea is to specify "02???" as the folder inside "011", and "???" as the file to be played once.

Is this command supported?
If yes, could anyone share an example of the correct syntax or usage?

Thank you!

**Comment by rnrobles** [comment]
@budimanaabuu  There are some boards with the firmware AT+VER1.0-20200911V02 that did not yet have the use of AT+AJ implemented, and others a little newer with the AT+VER2.3, when you turn on the board by serial port it tells you the version, I tried to update the board with the updata.bfu but I never succeeded, I was able to make a full backup of the .bin, I already asked for a 2.3 to be able to replace the firmware, but check what version you have

**Comment by VIYASHIKA04** [comment]
> Has anyone ever successfully used the AT command AT+AJ/01*/011_11.???? I would like to specify "01???" as the folder inside "011_11", and "???" as the file to be played once.
> 
> Similarly, using AT+AJ/02*/011*.???, the idea is to specify "02???" as the folder inside "011", and "???" as the file to be played once.
> 
> Is this command supported? If yes, could anyone share an example of the correct syntax or usage?
> 
> Thank you!


we are also tried  AT+AJ/02*/011*.??? command in bt 201 but it is not working .have you tried any other way to implement and also give us the link of v2.4 version released in 2021.if you know please let us know.
thanks

**Comment by wendylacosta** [comment]
Hello everyone, I'm using BT201 as a mp3 player with wired earphones. But I finding issues for my files. Basically, I have 12 songs in .flac that I put into TF card via file manager android. But BT201 only read 5 out of 12 songs. When I check  the other 7 songs in the TF card, it's corrupted I think? because it can't play songs when I put it in my phones.

Currently, I set the issues as "the phone to TF card transfering  get lost somehow." I want to ask you guys opininon about the issues please.

**Comment by wendylacosta** [comment]
> Hello everyone, I'm using BT201 as a mp3 player with wired earphones. But I finding issues for my files. Basically, I have 12 songs in .flac that I put into TF card via file manager android. But BT201 only read 5 out of 12 songs. When I check  the other 7 songs in the TF card, it's corrupted I think? because it can't play songs when I put it in my phones.
> 
> Currently, I set the issues as "the phone to TF card transfering  get lost somehow." I want to ask you guys opininon about the issues please.

Update: I changed the broken flac songs into mp3 as a temporary solution until I get my microsd reader. 🫡

**Comment by iplaythesax** [comment]
I'm using a ESP32 Wroom 32 in conjunction with the BT201 with the following code
```
#include <Arduino.h>

#define RXD2 16
#define TXD2 17

void setup() {
  Serial.begin(115200);
  while (!Serial) {
    delay(10);
  }

  Serial2.begin(115200, SERIAL_8N1, RXD2, TXD2);

  Serial.println("Serial2 started on RX=16, TX=17");
}

void loop() {
  if (Serial2.available()) {
    String incoming = Serial2.readStringUntil('\n');
    Serial.print("From Serial2: ");
    Serial.println(incoming);
  }

  if (Serial.available()) {
    String outgoing = Serial.readStringUntil('\n');
    Serial2.print(outgoing);
    Serial2.print('\n');
  }
}
```
The chip is running on version 2.3 updated 2019 11/15 
I have on the micro sd card folders set up like this
"01 folder"
→"01".mp3

"02 folder"
→"02".mp3

"03 folder"
→"03".mp3

"04 folder"
→"04 01 folder".
→→"04 01".mp3
→"04 02 folder"
→→"04 02".mp3

I have confirmed commands like
- AT+CA to AT+CW
However function AT+AF and AT+AJ don't seem to work and instead plays "02".mp3 on repeat after using the function once despite any parameters then "01".mp3 

Anyone have a reasoning to why this happens or can help fix it?

**Comment by revalogics** [comment]
> My BT201 on the V2.4 firmware. So technically, nothing had changed. All the serial commands are still the same, and everything else. 31 levels of volume (0 to 30) aren't the problem, guys, It still goes 0 to 100%. They made it in 31 levels so users don't have to click so many times just to adjust the volume. 0 is 0%, 30 is 100%, get it ?
> 
> If you're making a bluetooth speaker with B201, I have a few advices :
> 
> 1. Don't use the onboard amp, obviously, it's suck, it's a single channel, it's only 3-5W, connect the onboard 3.5mm to some amplifier board with some cables.
> 2. If you want to add the AUX input feature, use an audio relay, switch between BT201 3.5mm and AUX input to the amp
> 3. Search for AT+CK in the PDF, turn on bluetooth running in background, it'll be useful (more on this below)
> 4. When your BT201 already turned on bluetooth in backgound, it'll still send feedback about bluetooth status to serial while in other modes like USB/TF, PC sound card. You can take advanted of this, put some audio files in an SD card and use AT+AF, AT+AJ (READ THE PDF !!!) to play the these audio files as Prompt Tones (AT+CN00 to turn off onboard tones)
> 
> I don't have much free time, but will make a complete guide on this module when the project is done.
> 
> ![image](https://github.com/user-attachments/assets/e58f6f83-9029-45a7-92ab-aeb10c3b9496)

Hi @PhuocBull . I recently got a BT201 board, tried the AT Commands, but some of the commands doesn't work on my board (AT+AC00 to AT+AC04, about track repeats). I also don't know how to change the prompt voices even though I prepared the MP3 files to replace the prompts. I also noticed that your image here shows a V2.3 while mine seemed to be stuck at V1.0. Do you still have the firmware (updata.bfu) of this BT201 Board? I can't find the download link on the link you previously provided.

Many thanks.

<img width="812" height="568" alt="Image" src="https://github.com/user-attachments/assets/db0dfac1-ab79-47e7-b449-89f971929fdb" />

**Comment by revalogics** [comment]
> [@PhuocBull](https://github.com/PhuocBull) Hello do you know that can we put a limit of the top volume level ?

I also noticed the sound from the built-in amplifier on the board to be distorted at high volumes. I figured that the amplifier chip receives signal that is too loud so I replaced R1 and R2 from 22kΩ to 100kΩ to lower the output to tolerable levels.

This modification doesn't change the output level of the headphone out since it connects via a separate set of resistors (2x 10Ω, labeled as R7 and R8).

<img width="2173" height="1280" alt="Image" src="https://github.com/user-attachments/assets/1db44265-0735-4202-a4e1-e066a493fe8a" />
